pkg/apiserver/filters/devops.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2020 The KubeSphere Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	"kubesphere.io/kubesphere/pkg/simple/client/devops/jenkins"
+
+	"k8s.io/apimachinery/pkg/util/proxy"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/klog"
+
+	"kubesphere.io/kubesphere/pkg/apiserver/request"
+	"kubesphere.io/kubesphere/pkg/server/errors"
+)
+
+// WithDevOpsAPIServer proxy request to DevOps service if requests path has the APIGroup with devops.kubesphere.io
+func WithDevOpsAPIServer(handler http.Handler, config *jenkins.Options, failed proxy.ErrorResponder) http.Handler {
```

Comment:
For those business specific logic, it's better to use proxy, for example https://github.com/kubesphere/kubesphere/blob/master/pkg/kapis/notification/v1/register.go

Reply:
It's a really good example. But we will replace the proxy with the KubeSphere plugin mechanism which is under the testing stage. I believe it will be very soon.

/cc @wansir

Reply:
Before that, we'd better follow [this example](https://github.com/kubesphere/kubesphere/blob/master/pkg/kapis/notification/v1/register.go.) register proxy API endpoint. The dynamic  API registration is a new feature of 4.0, which will develop in a new branch.

Reply:
> Before that, we'd better follow [this example](https://github.com/kubesphere/kubesphere/blob/master/pkg/kapis/notification/v1/register.go.) register proxy API endpoint. The dynamic API registration is a new feature of 4.0, which will develop in a new branch.

The broken link `this example` has one more period.